File: app/src/main/java/me/ycdev/android/devtools/sampler/cpu/AppCpuStat.java
package me.ycdev.android.devtools.sampler.cpu;

import android.util.SparseArray;

public class AppCpuStat {
    public String pkgName;
    public SparseArray<ProcCpuStat> procSetStats = new SparseArray<>();

    public AppCpuStat(String pkgName) {
        this.pkgName = pkgName;
    }

    public long getTotal() {
        long total = 0;
        final int N = procSetStats.size();
        for (int i = 0; i < N; i++) {
            ProcCpuStat pidStat = procSetStats.valueAt(i);
            total += pidStat.getTotal();
        }
        return total;
    }

    public static AppCpuStat computeUsage(AppCpuStat oldStat, AppCpuStat newStat) {
        if (!oldStat.pkgName.equals(newStat.pkgName)) {
            return null;
        }
        AppCpuStat usage = new AppCpuStat(newStat.pkgName);
        final int N = newStat.procSetStats.size();
        for (int i = 0; i < N; i++) {
            int pid = newStat.procSetStats.keyAt(i);
            ProcCpuStat newPidStat = newStat.procSetStats.valueAt(i);
            ProcCpuStat oldPidStat = oldStat.procSetStats.get(pid);
            if (oldPidStat != null) {
                ProcCpuStat pidUsage = ProcCpuStat.computeUsage(oldPidStat, newPidStat);
                if (pidUsage != null) {
                    usage.procSetStats.put(pidUsage.pid, pidUsage);
                }
            }
        }
        return usage;
    }
}

File: app/src/main/java/me/ycdev/android/devtools/sampler/cpu/AppCpuStat.kt
package me.ycdev.android.devtools.sampler.cpu

import android.util.SparseArray

class AppCpuStat(var pkgName: String) {
    var procSetStats = SparseArray<ProcCpuStat?>()

    val total: Long
        get() {
            var total: Long = 0
            val size = procSetStats.size()
            for (i in 0 until size) {
                val pidStat = procSetStats.valueAt(i)
                total += pidStat?.total ?: 0
            }
            return total
        }

    companion object {
        fun computeUsage(oldStat: AppCpuStat, newStat: AppCpuStat): AppCpuStat {
            if (oldStat.pkgName != newStat.pkgName) {
                throw RuntimeException("Not same pkgName")
            }
            val usage = AppCpuStat(newStat.pkgName)
            val size = newStat.procSetStats.size()
            for (i in 0 until size) {
                val pid = newStat.procSetStats.keyAt(i)
                val newPidStat = newStat.procSetStats.valueAt(i)
                val oldPidStat = oldStat.procSetStats[pid]
                if (oldPidStat != null) {
                    val pidUsage: ProcCpuStat? =
                        ProcCpuStat.computeUsage(oldPidStat, newPidStat)
                    if (pidUsage != null) {
                        usage.procSetStats.put(pidUsage.pid, pidUsage)
                    }
                }
            }
            return usage
        }
    }
}